sysl2/sysl/ints.go

```diff
@@ -0,0 +1,215 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/anz-bank/sysl/src/proto"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type IntegrationParam struct {
+	m            *sysl.Module
+	excludes     StrSet
+	passthroughs StrSet
+	commonEndpts StrSet
+	drawableApps StrSet
+	apps         StrSet
+	deps         map[string]*AppDependency
+}
+
+func MakeIntegrationParam(mod *sysl.Module, excludes, passthroughs, highlights StrSet) *IntegrationParam {
+	return &IntegrationParam{
+		m:            mod,
+		excludes:     excludes,
+		passthroughs: passthroughs,
+		commonEndpts: MakeStrSet(".. * <- *", "*"),
+		drawableApps: highlights,
+		apps:         StrSet{},
+		deps:         map[string]*AppDependency{},
+	}
+}
+
+func (v *IntegrationParam) collectApplicationDependencies(apps StrSet) {
+	for app := range apps {
+		apps := v.m.GetApps()
+		for epname, endpt := range apps[app].GetEndpoints() {
+			v.handleStatement(app, epname, endpt.GetStmt())
+		}
+	}
+}
+
+func (v *IntegrationParam) handleStatement(appname, epname string, stmts []*sysl.Statement) {
+	for _, stmt := range stmts {
+		switch t := stmt.GetStmt().(type) {
+		case *sysl.Statement_Call:
+			targetName := getAppName(t.Call.GetTarget())
+			if added := v.addAppDependency(appname, epname, targetName, t.Call.Endpoint, stmt); added {
+				if t.Call.Endpoint != ".. * <- *" {
+					for epname, endpt := range v.m.GetApps()[targetName].GetEndpoints() {
+						v.handleStatement(targetName, epname, endpt.GetStmt())
+					}
+				}
+			}
+		case *sysl.Statement_Action, *sysl.Statement_Ret:
+			continue
+		case *sysl.Statement_Cond:
+			v.handleStatement(appname, epname, t.Cond.GetStmt())
+		case *sysl.Statement_Loop:
+			v.handleStatement(appname, epname, t.Loop.GetStmt())
+		case *sysl.Statement_LoopN:
+			v.handleStatement(appname, epname, t.LoopN.GetStmt())
+		case *sysl.Statement_Foreach:
+			v.handleStatement(appname, epname, t.Foreach.GetStmt())
+		case *sysl.Statement_Group:
+			v.handleStatement(appname, epname, t.Group.GetStmt())
+		case *sysl.Statement_Alt:
+			for _, choice := range t.Alt.GetChoice() {
+				v.handleStatement(appname, epname, choice.GetStmt())
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func (v *IntegrationParam) addAppDependency(sourceApp, sourceEndpt, targetApp, targetEndpt string, stmt *sysl.Statement) bool {
+	if len(v.apps) == 0 {
+		v.apps = v.apps.Union(v.drawableApps)
```

Comment:
this should move to initialization, right?

Reply:
yep. moved.